<!doctype html><html lang="en"><head><title>LisAln (LIberal Sequence ALigNment)</title><meta name="title" content="LisAln (LIberal Sequence ALigNment)"><meta name="description" content="LisAln (LIberal Sequence ALigNment) is a hit-the-button website to find orthologs (same genes/proteins across species) and paralogs (similar genes/proteins inside human) of a protein for general scientists or amateurs. It is developed at City of Hope National Medical Center. Users can either input a protein name, fasta sequence, or fasta file to search for a protein. Users can also select the sequence range of human protein and name type for species (scientific or general names) to display the alignment results from EMBL Clustal Omega and Uniprot. LisAln will automatically search the orthologs and paralogs from NCBI Landmark and Homologene databases. It will pick the appropriate proteins to display the results of whole sequence alignments, local aligned sequences that users are interested in, phylogenetic trees, sequence identity heatmap, etc. "><meta charset="utf-8"/><link rel="icon" href="/emoji.png"/><meta name="viewport" content="width=device-width,initial-scale=1"/><meta name="theme-color" content="#000000"/><meta name="language" content="English"><meta name="robots" content="index, follow"><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><meta name="keywords" content="Sequence Alignment, Paralogs, Orthologs, City of Hope, Protein, Fasta, Uniport, EMBL Clustal Omega, protein alignment, gene alignment, NCBI, NCBI Landmark, Phylogenetic trees"><meta name="revisit-after" content="30 days"><meta name="author" content="Daniel Ji, Maggie Li, Sandra Li, Binghui Shen, Hongzhi Li"><link rel="manifest" href="/manifest.json"/><meta property="og:type" content="website"><meta property="og:url" content="https://www.lisaln.org/"><meta property="og:title" content="LisAln (LIberal Sequence ALigNment)"><meta property="og:description" content="LisAln (LIberal Sequence ALigNment) is a hit-the-button website to find orthologs (same genes/proteins across species) and paralogs (similar genes/proteins inside human) of a protein for general scientists or amateurs. It is developed at City of Hope National Medical Center. Users can either input a protein name, fasta sequence, or fasta file to search for a protein. Users can also select the sequence range of human protein and name type for species (scientific or general names) to display the alignment results from EMBL Clustal Omega and Uniprot. LisAln will automatically search the orthologs and paralogs from NCBI Landmark and Homologene databases. It will pick the appropriate proteins to display the results of whole sequence alignments, local aligned sequences that users are interested in, phylogenetic trees, sequence identity heatmap, etc. 
    "><meta property="og:image" content="https://www.lisaln.org/preview.png"><meta property="twitter:card" content="https://www.lisaln.org/preview.png"><meta property="twitter:url" content="https://www.lisaln.org/"><meta property="twitter:title" content="LisAln (LIberal Sequence ALigNment)"><meta property="twitter:description" content="LisAln (LIberal Sequence ALigNment) is a hit-the-button website to find orthologs (same genes/proteins across species) and paralogs (similar genes/proteins inside human) of a protein for general scientists or amateurs. It is developed at City of Hope National Medical Center. Users can either input a protein name, fasta sequence, or fasta file to search for a protein. Users can also select the sequence range of human protein and name type for species (scientific or general names) to display the alignment results from EMBL Clustal Omega and Uniprot. LisAln will automatically search the orthologs and paralogs from NCBI Landmark and Homologene databases. It will pick the appropriate proteins to display the results of whole sequence alignments, local aligned sequences that users are interested in, phylogenetic trees, sequence identity heatmap, etc. 
    "><meta property="twitter:image" content="https://www.lisaln.org/preview.png"><link href="/static/css/main.e07274de.chunk.css" rel="stylesheet"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id="root"></div><script src="/static/js/runtime-main.c60b7f5c.js"></script><script src="/static/js/2.d1690937.chunk.js"></script><script src="/static/js/main.1f0248a9.chunk.js"></script></body></html>